### PR TITLE
[AWS REST Protocols] Implement AWS header (de)serialisation

### PR DESCRIPTION
**Pull request by Baccata**
Closes https://github.com/disneystreaming/smithy4s/issues/971
Closes https://github.com/disneystreaming/smithy4s/issues/970

Implement AWS Header de-serialisation, which abides by the following rules:  

* For each header name, the corresponding header values are merged together in a single string, with comma-separated values. 
* Values are quoted before being concatenated when the string contains a comma. 
* UNLESS the header values are timestamp values encoded using the `HTTP_DATE` encoding, in which case the values are not quoted (even if they contain commas). 

When encoding from a list, the serialisation logic therefore concatenates values together
When decoding into a list, the deserialisation logic multiple values from a single header. 

This PR also removes the concepts of `Total` and `Partial` metadata decoder, as the partiality is now handled at the schema level rather than the codec level. 

Although this logic is implemented in `core`, it is only wired in the AWS module.

### CODE REVIEW

**Comment by Baccata** [comment]
The number of failing tests for AWS Rest Json 1 is now down to 6 \o/

File: modules/core/test/src/smithy4s/http/internals/HeaderSplitSpec.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import munit._
+
+class HeaderSplitSpec extends FunSuite {
+
+  test("Header split: single value") {
+    checkNormal("abcd efgh")("abcd efgh")
+  }
+
+  test("Header split: multiple normal strings") {
+    checkNormal("a, b, c")("a", "b", "c")
```
Comment (daddykotex) [inline modules/core/test/src/smithy4s/http/internals/HeaderSplitSpec.scala:28]: maybe one test case with no space after the comma?

File: modules/core/src/smithy4s/http/Metadata.scala [hunk 2]
```diff
@@ -151,26 +150,13 @@ object Metadata {
...
 
   /**
-    * Decodes all the fields that are bound to http metadata,
-    * shoving values in a map.
+    * If possible, decode the data from fields that are bound to http metadata.
     */
-  def decodePartial[A](metadata: Metadata)(implicit
-      decoder: PartialDecoder[A]
-  ): Either[MetadataError, MetadataPartial[A]] =
+  def decode[A](metadata: Metadata)(implicit
```
Comment (daddykotex) [inline modules/core/src/smithy4s/http/Metadata.scala:155]: I'm surprised no calls to `Metadata.decode` happens outside of test files

Or am I missing something

Comment (Baccata) [reply modules/core/src/smithy4s/http/Metadata.scala:155]: @daddykotex yeah, it's rare to use `Metadata.decode` in this codebase as we use `Metadata.Decoder.fromSchema` instead.

**Review by lewisjkl — APPROVED** [review]

**Comment by daddykotex** [comment]
I see failure related to content encoding in here, I'll check this out and try to fix it

**Pull request merged**